File: packages/earthsci-toolkit/src/units-fixture.test.ts
/**
 * Units fixtures consumption runner (gt-dt0o).
 *
 * The three `units_*.esm` files in `tests/valid/` carry inline `tests`
 * blocks (id / parameter_overrides / initial_conditions / time_span /
 * assertions) added in gt-p3v. Schema parse coverage is asserted in
 * `units.test.ts`'s Cross-binding units fixtures suite. This file closes
 * the schema-vs-execution gap: every assertion's target (all of which
 * are observed variables at t = 0) is actually evaluated under the
 * test's bindings and compared against the expected value within the
 * resolved tolerance (assertion → test → model, falling back to
 * rtol = 1e-6).
 *
 * Corrupting an expected value in any fixture — or reverting the
 * `pressure_drop` fix from gt-p3v — must cause this suite to fail.
 */

import { describe, it, expect } from 'vitest'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { evaluate } from './expression.js'
import type { EsmFile, Expr, Model, ModelVariable, Test, Tolerance } from './types.js'

const FIXTURES_DIR = join(__dirname, '..', '..', '..', 'tests', 'valid')
const FIXTURES = [
  'units_conversions.esm',
  'units_dimensional_analysis.esm',
  'units_propagation.esm',
]

type AnyTol = { abs?: number; rel?: number } | undefined

function resolveTol(
  modelTol: AnyTol,
  testTol: AnyTol,
  assertionTol: AnyTol,
): { rel: number; abs: number } {
  for (const cand of [assertionTol, testTol, modelTol]) {
    if (cand === undefined || cand === null) continue
    return { rel: cand.rel ?? 0, abs: cand.abs ?? 0 }
  }
  return { rel: 1e-6, abs: 0 }
}

/**
 * Resolve every observed variable in `model` into `bindings` by iterated
 * substitution. The shared `evaluate` throws on the first unbound
 * variable; we swallow that as "dependencies not yet resolved" and
 * retry. Cycle-free fixtures converge in at most one pass per observed
 * variable.
 */
function resolveObserved(model: Model, bindings: Map<string, number>): void {
  const variables = model.variables ?? {}
  const entries = Object.entries(variables)
  for (let pass = 0; pass <= entries.length; pass++) {
    let progress = false
    for (const [vname, v] of entries) {
      const variable = v as ModelVariable
      if (variable.type !== 'observed') continue
      if (bindings.has(vname)) continue
      if (variable.expression === undefined) continue
      try {
        bindings.set(vname, evaluate(variable.expression as Expr, bindings))
        progress = true
      } catch (err) {
        if (err instanceof Error && err.message.startsWith('Unbound variable')) {
          continue
        }
        throw err
      }
    }
    if (!progress) return
  }
}

function assertWithTolerance(
  label: string,
  actual: number,
  expected: number,
  rel: number,
  abs: number,
): void {
  if (abs > 0 && expected === 0) {
    expect(Math.abs(actual - expected), label).toBeLessThanOrEqual(abs)
    return
  }
  if (rel > 0) {
    const bound = Math.max(rel * Math.max(Math.abs(expected), 1e-300), abs)
    expect(Math.abs(actual - expected), label).toBeLessThanOrEqual(bound)
    return
  }
  expect(Math.abs(actual - expected), label).toBeLessThanOrEqual(abs)
}

function buildBindings(model: Model, t: Test): Map<string, number> {
  const bindings = new Map<string, number>()
  for (const [vname, vraw] of Object.entries(model.variables ?? {})) {
    const variable = vraw as ModelVariable
    if ((variable.type === 'parameter' || variable.type === 'state')
        && typeof variable.default === 'number') {
      bindings.set(vname, variable.default)
    }
  }
  for (const [k, v] of Object.entries(t.initial_conditions ?? {})) {
    bindings.set(k, v as number)
  }
  for (const [k, v] of Object.entries(t.parameter_overrides ?? {})) {
    bindings.set(k, v as number)
  }
  return bindings
}

describe('Units fixtures inline tests execution (gt-dt0o)', () => {
  for (const fname of FIXTURES) {
    describe(fname, () => {
      const raw = readFileSync(join(FIXTURES_DIR, fname), 'utf8')
      const file = JSON.parse(raw) as EsmFile
      const models = file.models ?? {}
      const modelEntries = Object.entries(models)

      it('has at least one inline test across its models', () => {
        const count = modelEntries.reduce(
          (sum, [, m]) => sum + ((m as Model).tests?.length ?? 0),
          0,
        )
        expect(count).toBeGreaterThan(0)
      })

      for (const [mname, m] of modelEntries) {
        const model = m as Model
        for (const t of model.tests ?? []) {
          it(`${mname}/${t.id}`, () => {
            const bindings = buildBindings(model, t)
            resolveObserved(model, bindings)
            for (const a of t.assertions) {
              const { rel, abs } = resolveTol(
                model.tolerance as AnyTol,
                t.tolerance as AnyTol,
                a.tolerance as AnyTol,
              )
              expect(bindings.has(a.variable),
                `${fname}::${mname}::${t.id}: ${a.variable} not resolved`,
              ).toBe(true)
              assertWithTolerance(
                `${fname}::${mname}::${t.id}::${a.variable}`,
                bindings.get(a.variable)!,
                a.expected,
                rel,
                abs,
              )
            }
          })
        }
      }
    })
  }
})

File: packages/esm-format-go/pkg/esm/units_fixture_test.go
package esm

// Units fixtures consumption runner (gt-dt0o).
//
// The three units_*.esm files in tests/valid/ carry inline `tests` blocks
// (id / parameter_overrides / initial_conditions / time_span / assertions)
// added in gt-p3v. Schema parse coverage is asserted in
// units_fixtures_test.go (TestUnitsFixturesCrossBinding). This file closes
// the schema-vs-execution gap: every assertion's target (all of which are
// observed variables at t = 0) is actually evaluated under the test's
// bindings and compared against the expected value within the resolved
// tolerance (assertion → test → model, falling back to rtol = 1e-6).
//
// Corrupting an expected value in any fixture — or reverting the
// pressure_drop fix from gt-p3v — must cause this suite to fail.
//
// The Go `Model` struct does not currently carry the inline `tests` block,
// so this runner walks the raw JSON directly rather than going through
// LoadString. The shape is schema-validated elsewhere; here we only need
// the fields listed above.

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"testing"
)

type unitsFixtureTolerance struct {
	Rel float64
	Abs float64
}

func parseTol(raw interface{}) (unitsFixtureTolerance, bool) {
	if raw == nil {
		return unitsFixtureTolerance{}, false
	}
	m, ok := raw.(map[string]interface{})
	if !ok {
		return unitsFixtureTolerance{}, false
	}
	out := unitsFixtureTolerance{}
	if v, ok := m["rel"]; ok {
		if f, ok := v.(float64); ok {
			out.Rel = f
		}
	}
	if v, ok := m["abs"]; ok {
		if f, ok := v.(float64); ok {
			out.Abs = f
		}
	}
	return out, true
}

func resolveUnitsFixtureTol(modelTol, testTol, assertionTol interface{}) (float64, float64) {
	for _, raw := range []interface{}{assertionTol, testTol, modelTol} {
		if t, ok := parseTol(raw); ok {
			return t.Rel, t.Abs
		}
	}
	return 1e-6, 0.0
}

// evalFixtureExpr evaluates a raw JSON expression (numbers, variable
// strings, operator nodes as map[string]interface{}). Returns ok=false if
// any variable reference is unbound so the caller can defer observed
// resolution.
func evalFixtureExpr(expr interface{}, bindings map[string]float64) (float64, bool) {
	switch e := expr.(type) {
	case nil:
		return 0, false
	case float64:
		return e, true
	case int:
		return float64(e), true
	case string:
		v, ok := bindings[e]
		return v, ok
	case map[string]interface{}:
		opRaw, ok := e["op"].(string)
		if !ok {
			return 0, false
		}
		argsRaw, ok := e["args"].([]interface{})
		if !ok {
			return 0, false
		}
		args := make([]float64, len(argsRaw))
		for i, a := range argsRaw {
			v, ok := evalFixtureExpr(a, bindings)
			if !ok {
				return 0, false
			}
			args[i] = v
		}
		switch opRaw {
		case "+":
			sum := 0.0
			for _, v := range args {
				sum += v
			}
			return sum, true
		case "-":
			switch len(args) {
			case 1:
				return -args[0], true
			case 2:
				return args[0] - args[1], true
			default:
				panic(fmt.Sprintf("'-' needs 1 or 2 args, got %d", len(args)))
			}
		case "*":
			prod := 1.0
			for _, v := range args {
				prod *= v
			}
			return prod, true
		case "/":
			if len(args) != 2 {
				panic(fmt.Sprintf("'/' needs 2 args, got %d", len(args)))
			}
			return args[0] / args[1], true
		case "^", "**":
			if len(args) != 2 {
				panic(fmt.Sprintf("'^' needs 2 args, got %d", len(args)))
			}
			return math.Pow(args[0], args[1]), true
		case "log":
			return math.Log(args[0]), true
		case "exp":
			return math.Exp(args[0]), true
		case "sqrt":
			return math.Sqrt(args[0]), true
		case "sin":
			return math.Sin(args[0]), true
		case "cos":
			return math.Cos(args[0]), true
		case "tan":
			return math.Tan(args[0]), true
		case "abs":
			return math.Abs(args[0]), true
		default:
			panic(fmt.Sprintf("unsupported op: %q", opRaw))
		}
	default:
		return 0, false
	}
}

func resolveUnitsFixtureObserved(variables map[string]interface{}, bindings map[string]float64) {
	n := len(variables) + 1
	for i := 0; i < n; i++ {
		progress := false
		for vname, vraw := range variables {
			v, ok := vraw.(map[string]interface{})
			if !ok {
				continue
			}
			if v["type"] != "observed" {
				continue
			}
			if _, already := bindings[vname]; already {
				continue
			}
			expr, has := v["expression"]
			if !has {
				continue
			}
			val, ok := evalFixtureExpr(expr, bindings)
			if !ok {
				continue
			}
			bindings[vname] = val
			progress = true
		}
		if !progress {
			return
		}
	}
}

func buildUnitsFixtureBindings(
	variables map[string]interface{},
	initialConditions, parameterOverrides map[string]interface{},
) map[string]float64 {
	bindings := map[string]float64{}
	for vname, vraw := range variables {
		v, ok := vraw.(map[string]interface{})
		if !ok {
			continue
		}
		typ, _ := v["type"].(string)
		if typ != "parameter" && typ != "state" {
			continue
		}
		if d, ok := v["default"].(float64); ok {
			bindings[vname] = d
		}
	}
	for k, v := range initialConditions {
		if f, ok := v.(float64); ok {
			bindings[k] = f
		}
	}
	for k, v := range parameterOverrides {
		if f, ok := v.(float64); ok {
			bindings[k] = f
		}
	}
	return bindings
}

func mapOr(m map[string]interface{}, key string) map[string]interface{} {
	if v, ok := m[key].(map[string]interface{}); ok {
		return v
	}
	return nil
}

func checkUnitsFixtureAssertion(
	t *testing.T,
	label string,
	actual, expected, rel, abs float64,
) {
	t.Helper()
	diff := math.Abs(actual - expected)
	bound := abs
	if rel > 0 {
		relBound := rel * math.Max(math.Abs(expected), math.SmallestNonzeroFloat64)
		if relBound > bound {
			bound = relBound
		}
	}
	if !(diff <= bound) {
		t.Fatalf("%s: actual=%g expected=%g rel=%g abs=%g diff=%g",
			label, actual, expected, rel, abs, diff)
	}
}

func TestUnitsFixturesInlineTestsExecution(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}
	fixtures := []string{
		"units_conversions.esm",
		"units_dimensional_analysis.esm",
		"units_propagation.esm",
	}
	totalTests := 0
	for _, name := range fixtures {
		t.Run(name, func(t *testing.T) {
			path := filepath.Join(repoRoot, "tests", "valid", name)
			raw, err := os.ReadFile(path)
			if err != nil {
				t.Fatalf("read %s: %v", path, err)
			}
			var doc map[string]interface{}
			if err := json.Unmarshal(raw, &doc); err != nil {
				t.Fatalf("unmarshal %s: %v", name, err)
			}
			models := mapOr(doc, "models")
			if len(models) == 0 {
				t.Fatalf("%s: no models", name)
			}
			fixtureTests := 0
			for mname, mraw := range models {
				model, ok := mraw.(map[string]interface{})
				if !ok {
					continue
				}
				testsRaw, _ := model["tests"].([]interface{})
				if len(testsRaw) == 0 {
					continue
				}
				variables := mapOr(model, "variables")
				modelTol := model["tolerance"]
				for _, tcRaw := range testsRaw {
					tc, ok := tcRaw.(map[string]interface{})
					if !ok {
						continue
					}
					id, _ := tc["id"].(string)
					ic := mapOr(tc, "initial_conditions")
					po := mapOr(tc, "parameter_overrides")
					bindings := buildUnitsFixtureBindings(variables, ic, po)
					resolveUnitsFixtureObserved(variables, bindings)

					assertionsRaw, _ := tc["assertions"].([]interface{})
					testTol := tc["tolerance"]
					subname := fmt.Sprintf("%s/%s", mname, id)
					t.Run(subname, func(t *testing.T) {
						for _, aRaw := range assertionsRaw {
							a, ok := aRaw.(map[string]interface{})
							if !ok {
								continue
							}
							variable, _ := a["variable"].(string)
							expected, _ := a["expected"].(float64)
							aTol := a["tolerance"]
							rel, abs := resolveUnitsFixtureTol(modelTol, testTol, aTol)
							actual, ok := bindings[variable]
							if !ok {
								t.Fatalf("%s::%s::%s: observed %q did not resolve (bindings=%v)",
									name, mname, id, variable, bindings)
							}
							label := fmt.Sprintf("%s::%s::%s::%s", name, mname, id, variable)
							checkUnitsFixtureAssertion(t, label, actual, expected, rel, abs)
						}
					})
					fixtureTests++
					totalTests++
				}
			}
			if fixtureTests == 0 {
				t.Fatalf("%s: expected at least one inline test across its models", name)
			}
		})
	}
	if totalTests == 0 {
		t.Fatalf("expected at least one inline test across the units fixtures")
	}
}
